### PR TITLE
Document the v3 notifier endpoint

### PR DESCRIPTION
**Pull request by zefer**
I noticed that we have some contradictions and inconsistencies in our
existing documentation for our notifier endpoints. This attempts to fix
that.

Reviewers: please pay particular attention to the accuracy of the POST
Data Fields section.

### CODE REVIEW

File: source/index.md [hunk 1]
```diff
@@ -143,6 +143,209 @@ collectionName | `projects` | Each API has different collection name. Some APIs
...
+
+`POST https://airbrake.io/api/v3/projects/PROJECT_ID/notices?key=PROJECT_KEY`
+
+```shell
+curl -X POST -H "Content-Type: application/json" -d '{ ... }' "https://airbrake.io/api/v3/projects/PROJECT_ID/notices?key=PROJECT_KEY"
+```
+
+### POST data
+
+Incoming errors should be described using JSON in the POST body of the request.
```
Comment (dvdplm) [inline source/index.md]: "Send new errors to airbrake by POSTing a JSON encoded body"?

"Incoming" – it's incoming to us, but outgoing for them (the reader), so I'd avoid it.

File: source/index.md [hunk 2]
```diff
@@ -143,6 +143,209 @@ collectionName | `projects` | Each API has different collection name. Some APIs
...
+
+On Success, the API returns a `201 Created` status with the following JSON data.
+
+Field | Comment
+----- | -------
+id | The ID of the newly create error notice. This can be used to [query the status of this error notice](#show-notice-status-v4)
+url | A URL that will take you to the error on the Airbrake dashboard
+
+**Note**: a success response means that the data has been received and will be
+processed very shortly. Use the `url` or `id` in the response to query the
```
Comment (dvdplm) [inline source/index.md]: "… data has been received and accepted for processing. This normally happens within a second or two."?

File: source/index.md [hunk 3]
```diff
@@ -143,6 +143,209 @@ collectionName | `projects` | Each API has different collection name. Some APIs
...
+status of an error. This will tell you if the error has been processed, or if it
+has been rejected for reasons such as invalid JSON.
```
Comment (dvdplm) [inline source/index.md]: @vmihailenco what other reasons would there be for rejection? Rate limiting?

File: source/index.md [hunk 4]
```diff
@@ -143,6 +143,209 @@ collectionName | `projects` | Each API has different collection name. Some APIs
...
+errors | true | An array of objects describing the error that occurred
+errors/{i}/class | true | The class name or type of error that occurred
+errors/{i}/message | false | A short message describing the error that occurred
+errors/{i}/backtrace | true | An array of objects describing each line of the error's backtrace
+errors/{i}/backtrace/{i}/file | true | The full path of the file in this entry of the backtrace
+errors/{i}/backtrace/{i}/line | false | The file's line number in this entry of the backtrace
+errors/{i}/backtrace/{i}/function | false | When available, the function or method name in this entry of the backtrace
+context | false | An object describing additional context for this error
+context/environment | false | The name of the server environment in which the error occurred, e.g. "staging", "production", etc
+context/component | false | The component in which the error occurred. In MVC frameworks like Rails, this should be set to the controller. Otherwise, this can be set to a route or other request category
```
Comment (dvdplm) [inline source/index.md]: "The component or module…"?

File: source/index.md [hunk 5]
```diff
@@ -143,6 +143,209 @@ collectionName | `projects` | Each API has different collection name. Some APIs
...
+context/action | false | The action in which the error occurred. If each request is routed to a controller action, this should be set here. Otherwise, this can be set to a method or other request subcategory
+context/os | false | Details of the operating system on which the error occurred
+context/language | false | Describe the language on which the error occurred, e.g. "Ruby 2.1.1"
+context/version | false | Describe the application version, e.g. "v1.2.3"
+context/url | false | The application's URL
+context/userAgent | false | The requesting browser's full user-agent string
+context/rootDirectory | false | The application's root directory path
+context/userId | false | If applicable, the current user's ID
+context/userName | false | If applicable, the current user's username
+context/userEmail | false | If applicable, the current user's email address
```
Comment (dvdplm) [inline source/index.md]: I believe we actually allow different attributes to be sent here, but @vmihailenco has the details.

File: source/index.md [hunk 6]
```diff
@@ -143,6 +143,209 @@ collectionName | `projects` | Each API has different collection name. Some APIs
...
+## POST Data Fields v3
+
+Field | Required | Description
+------|----------|------------
+notifier | true | An object describing the notifier client library
+notifier/name | true | The name of the notifier client submitting the request, e.g. "airbrake-js"
+notifier/version | true | The version number of the notifier client submitting the request, e.g. "1.2.3"
+notifier/url | true | A URL at which more information can be obtained concerning the notifier client
+errors | true | An array of objects describing the error that occurred
+errors/{i}/class | true | The class name or type of error that occurred
```
Comment (dvdplm) [inline source/index.md]: I wonder, should we mention somewhere that while the api support multiple errors to be sent, our UI does not (nor does the ruby notifier atm)?

**Comment by dvdplm** [comment]
Superb work here @zefer, much appreciated.

File: source/index.md (hunk 3 above)
Comment (vmihailenco) [inline source/index.md]: Rate limiting and error from old version.

File: source/index.md [hunk 7]
```diff
@@ -143,6 +143,209 @@ collectionName | `projects` | Each API has different collection name. Some APIs
 start | abcdefg | Position of the first element in the result set.
 end | abcdefg | Position of the last element in the result set.
 
+# Error Notification v3
+
+## Notify Error v3
```
Comment (vmihailenco) [inline source/index.md]: The rest of docs uses "Create notice v3" pattern. I don't claim that it is better, but can we be consistent?

File: source/index.md [hunk 8]
```diff
@@ -143,6 +143,209 @@ collectionName | `projects` | Each API has different collection name. Some APIs
...
+
+Notify Airbrake that a new error has occurred in your application.
```
Comment (vmihailenco) [inline source/index.md]: Probably "The API notifies Airbrake ..."

File: source/index.md (hunk 1 above)
Comment (vmihailenco) [inline source/index.md]: https://airbrake.io/docs/#create-deploy-v4 says "The API expects JSON data.". Can we find consistent wording?

File: source/index.md [hunk 9]
```diff
@@ -143,6 +143,209 @@ collectionName | `projects` | Each API has different collection name. Some APIs
...
+
+### POST data
+
+Incoming errors should be described using JSON in the POST body of the request.
+
+See [POST Data Fields](#post-data-fields-v3),
+[POST Data Example](#post-data-example-v3) &
+[POST Data Schema](#post-data-schema-v3).
+
+### Response
```
Comment (vmihailenco) [inline source/index.md]: I would place this after `#post-data-schema-v3` so we describe request before response.

File: source/index.md [hunk 10]
```diff
@@ -143,6 +143,209 @@ collectionName | `projects` | Each API has different collection name. Some APIs
...
+url | A URL that will take you to the error on the Airbrake dashboard
+
+**Note**: a success response means that the data has been received and will be
+processed very shortly. Use the `url` or `id` in the response to query the
+status of an error. This will tell you if the error has been processed, or if it
+has been rejected for reasons such as invalid JSON.
+
+## POST Data Fields v3
+
+Field | Required | Description
```
Comment (vmihailenco) [inline source/index.md]: This format is not very readable IMHO. Can we use JSON with comments here, e.g.:

```
{
  notifier: {
    // The name of the notifier client submitting the request, e.g. "airbrake-js"
    name: "airbrake-js", 
    // The version number of the notifier client submitting the request, e.g. "1.2.3"
    version: "1.2.3"
  }
}
```

Probably we can do in a separate PR.

Comment (zefer) [inline source/index.md]: Good idea. Ill try this out in a separate PR and will focus on getting the content right in this PR.

File: source/index.md (hunk 3 above)
Comment (zefer) [inline source/index.md]: Thanks. Could you please explain what you mean by error from old version?

Comment (vmihailenco) [inline source/index.md]: If notice.context.version < project.min_app_version we reject the notice.

File: source/index.md (hunk 6 above)
Comment (vmihailenco) [inline source/index.md]: class -> type

File: source/index.md [hunk 11]
```diff
@@ -143,6 +143,209 @@ collectionName | `projects` | Each API has different collection name. Some APIs
...
+notifier | true | An object describing the notifier client library
+notifier/name | true | The name of the notifier client submitting the request, e.g. "airbrake-js"
+notifier/version | true | The version number of the notifier client submitting the request, e.g. "1.2.3"
+notifier/url | true | A URL at which more information can be obtained concerning the notifier client
+errors | true | An array of objects describing the error that occurred
+errors/{i}/class | true | The class name or type of error that occurred
+errors/{i}/message | false | A short message describing the error that occurred
+errors/{i}/backtrace | true | An array of objects describing each line of the error's backtrace
+errors/{i}/backtrace/{i}/file | true | The full path of the file in this entry of the backtrace
+errors/{i}/backtrace/{i}/line | false | The file's line number in this entry of the backtrace
```
Comment (vmihailenco) [inline source/index.md]: Add column - file:line:column

File: source/index.md (hunk 3 above)
Comment (dvdplm) [inline source/index.md]: We don't have any UI to set the `min_app_version` though. If min_app_version is not set  I guess we don't reject?

Comment (dvdplm) [inline source/index.md]: The comparison, is it "smart" or simple alphabetic string comparison? I.e. is "v1.2.3-alpha" > "v1.2.3-beta"?

File: source/index.md (hunk 11 above)
Comment (zefer) [inline source/index.md]: `errors/{i}/backtrace/{i}/column`?

File: source/index.md (hunk 3 above)
Comment (zefer) [inline source/index.md]: I think this probably isn't the place to document the `min_app_version` feature.

**Comment by zefer** [comment]
Updated. PTAL.

File: source/index.md (hunk 3 above)
Comment (vmihailenco) [inline source/index.md]: You can set min app version at
https://getexceptional.airbrake.io/projects/109058/edit?project_id=109058

Comment (vmihailenco) [inline source/index.md]: It is dumb. We should start using https://github.com/blang/semver

File: source/index.md [hunk 12]
```diff
@@ -143,6 +143,210 @@ collectionName | `projects` | Each API has different collection name. Some APIs
...
+errors | true | An array of objects describing the error that occurred
+errors/{i}/type | true | The class name or type of error that occurred
+errors/{i}/message | false | A short message describing the error that occurred
+errors/{i}/backtrace | true | An array of objects describing each line of the error's backtrace
+errors/{i}/backtrace/{i}/file | true | The full path of the file in this entry of the backtrace
+errors/{i}/backtrace/{i}/line | false | The file's line number in this entry of the backtrace
+errors/{i}/backtrace/{i}/column | false | The line's column number in this entry of the backtrace
+errors/{i}/backtrace/{i}/function | false | When available, the function or method name in this entry of the backtrace
+context | false | An object describing additional context for this error
+context/environment | false | The name of the server environment in which the error occurred, e.g. "staging", "production", etc
```
Comment (vmihailenco) [inline source/index.md]: The full list of supported params is here: https://github.com/airbrake/goab/blob/master/models/notice.go#L98 . We should document that code and copy comments here.

File: source/index.md [hunk 13]
```diff
@@ -143,6 +143,210 @@ collectionName | `projects` | Each API has different collection name. Some APIs
...
+    "environment": {"type": "object"},
+    "session": {"type": "object"},
+    "params": {"type": "object"}
+  }
+}
+```
+
+## Response
+
+On Success, the API returns a `201 Created` status with the following JSON data.
```
Comment (vmihailenco) [inline source/index.md]: Downcase Success?

File: source/index.md [hunk 14]
```diff
@@ -143,6 +143,210 @@ collectionName | `projects` | Each API has different collection name. Some APIs
...
+
+Field | Comment
+----- | -------
+id | The ID of the newly create error notice. This can be used to [query the status of this error notice](#show-notice-status-v4)
```
Comment (vmihailenco) [inline source/index.md]: newly created

Comment (vmihailenco) [inline source/index.md]: Dot in the end of sentence.

File: source/index.md [hunk 15]
```diff
@@ -143,6 +143,210 @@ collectionName | `projects` | Each API has different collection name. Some APIs
...
+url | A URL that will take you to the error on the Airbrake dashboard
```
Comment (vmihailenco) [inline source/index.md]: Dot.

File: source/index.md [hunk 16]
```diff
@@ -143,6 +143,210 @@ collectionName | `projects` | Each API has different collection name. Some APIs
...
+
+**Note**: a success response means that the data has been received and accepted
+for processing. Use the `url` or `id` in the response to query the status of an
```
Comment (vmihailenco) [inline source/index.md]: How you can use `url` to query the status? `url` is just a short link to the notice.

File: source/index.md [hunk 17]
```diff
@@ -143,6 +143,210 @@ collectionName | `projects` | Each API has different collection name. Some APIs
...
+
+### HTTP Request
+
+`POST https://airbrake.io/api/v3/projects/PROJECT_ID/notices?key=PROJECT_KEY`
+
+```shell
+curl -X POST -H "Content-Type: application/json" -d '{ ... }' "https://airbrake.io/api/v3/projects/PROJECT_ID/notices?key=PROJECT_KEY"
+```
+
+### POST data
```
Comment (vmihailenco) [inline source/index.md]: `Post Data` though the rest of the doc does not use titlecase. Do you plan to update it?

File: source/index.md [hunk 18]
```diff
@@ -143,6 +143,210 @@ collectionName | `projects` | Each API has different collection name. Some APIs
...
+
+The API expects JSON data.
+
+See [POST Data Fields](#post-data-fields-v3),
+[POST Data Example](#post-data-example-v3) &
+[POST Data Schema](#post-data-schema-v3).
+
+## POST Data Fields v3
```
Comment (vmihailenco) [inline source/index.md]: This should be h3?

File: source/index.md [hunk 19]
```diff
@@ -143,6 +143,210 @@ collectionName | `projects` | Each API has different collection name. Some APIs
...
+  },
+  "params": {
+    "page": "3",
+    "sort": "name",
+    "direction": "asc"
+  }
+}
+```
+
+## POST Data Schema v3
```
Comment (vmihailenco) [inline source/index.md]: h3 here and below.

File: source/index.md [hunk 20]
```diff
@@ -143,6 +143,210 @@ collectionName | `projects` | Each API has different collection name. Some APIs
...
+context/userAgent | false | The requesting browser's full user-agent string
+context/rootDirectory | false | The application's root directory path
+context/userId | false | If applicable, the current user's ID
+context/userName | false | If applicable, the current user's username
+context/userEmail | false | If applicable, the current user's email address
+environment | false | An object containing the current environment variables. Where the key is the variable name, e.g. `{ "PORT": "443", "CODE_NAME": "gorilla" }`
+session | false | An object containing the current session variables. Where the key is the variable name, e.g. `{ "basket_total": "1234", "user_id": "123" }`
+params | false | An object containing the request parameters. Where the key is the parameter name, e.g. `{ "page": "3", "sort": "desc" }`
+
+## POST Data Example v3
```
Comment (vmihailenco) [inline source/index.md]: Remove this section and move example to the `HTTP Request`.

File: source/index.md (hunk 17 above)
Comment (zefer) [inline source/index.md]: I see a mix, which do we prefer, and I'll standardise? I normally prefer normal case.

Comment (vmihailenco) [inline source/index.md]: Pick whatever you prefer, but please update whole doc.

File: source/index.md (hunk 16 above)
Comment (zefer) [inline source/index.md]: It also returns the status if the notice was rejected, for example:

```
curl -i "http://airbrake.io/locate/1374678579645155431"
HTTP/1.1 404 Not Found
Server: nginx
Content-Type: text/plain; charset=utf-8
Content-Length: 85
Vary: Accept-Encoding
Age: 9
Date: Thu, 12 Mar 2015 16:41:17 GMT
Connection: keep-alive

Notice 1374678579645155431 was rejected: invalid character ':' after top-level value.
```

**Comment by zefer** [comment]
PTAL

I'll standardise the title formatting in a separate PR after merging this one.

File: source/index.md [hunk 21]
```diff
@@ -143,6 +143,207 @@ collectionName | `projects` | Each API has different collection name. Some APIs
...
+  },
+  "params": {
+    "page": "3",
+    "sort": "name",
+    "direction": "asc"
+  }
+}
+```
+
+### POST Data
```
Comment (vmihailenco) [inline source/index.md]: Move this to the previous paragraph since it describes request and is small.

File: source/index.md [hunk 22]
```diff
@@ -143,6 +143,207 @@ collectionName | `projects` | Each API has different collection name. Some APIs
...
+
+The API expects JSON data.
+
+See [POST Data Fields](#post-data-fields-v3) &
+[POST Data Schema](#post-data-schema-v3).
+
+### POST Data Fields v3
```
Comment (vmihailenco) [inline source/index.md]: No need to repeat that this is v3. I would merge this into previous paragraph.

File: source/index.md [hunk 23]
```diff
@@ -143,6 +143,207 @@ collectionName | `projects` | Each API has different collection name. Some APIs
...
+context/userAgent | false | The requesting browser's full user-agent string.
+context/rootDirectory | false | The application's root directory path.
+context/userId | false | If applicable, the current user's ID.
+context/userName | false | If applicable, the current user's username.
+context/userEmail | false | If applicable, the current user's email address.
+environment | false | An object containing the current environment variables. Where the key is the variable name, e.g. `{ "PORT": "443", "CODE_NAME": "gorilla" }`.
+session | false | An object containing the current session variables. Where the key is the variable name, e.g. `{ "basket_total": "1234", "user_id": "123" }`.
+params | false | An object containing the request parameters. Where the key is the parameter name, e.g. `{ "page": "3", "sort": "desc" }`.
+
+### POST Data Schema v3
```
Comment (vmihailenco) [inline source/index.md]: Remove v3. Also This is JSON schema, not "post data schema"

File: source/index.md (hunk 22 above)
Comment (zefer) [inline source/index.md]: The previous paragraph has changed since this comment was made. Can you please clarify with the new context?

File: source/index.md (hunk 23 above)
Comment (zefer) [inline source/index.md]: The `v3` ensures we can link to this header

**Comment by vmihailenco** [comment]
LGTM. Great work.

I will send a separate PR with my suggestions.

**Comment by zefer** [comment]
Good plan. Thank you for the input.

File: source/index.md (hunk 8 above)
Comment (dvdplm) [inline source/index.md]: I'd avoid using "notify" all together. How about "Sends a new error to Airbrake"?

File: source/index.md (hunk 7 above)
Comment (dvdplm) [inline source/index.md]: I think we should be careful about using "notice" when communicating to the outside. It's not clear that "notice" == "error occurrence". I suggest "Send error v3".

Comment (vmihailenco) [inline source/index.md]: I have opinion on this too, but it is the opposite :) I think we should use
notice word, but add an explanation what it means, because we can't
completely hide it since `notice` is present it the URL and JSON fields
(group.notice_count, group.last_notice_id etc).

Comment (dvdplm) [inline source/index.md]: That's a possibility  too ofc, but means more work to communicate clearly with customers. 
Ideally we should change *all* uses of `notice` into `error`; in the meanwhile I'd stick with `error` as much as possible.
